Harden GenAPI

dougbu opened the pull request:
- dotnet/arcade#2066
- dotnet/arcade#3144 (kind of)
- dotnet/arcade#4488 except problem (5)
  - requires significant rewrite to look ahead or maintain sufficient context for `#nullable enable` gen
  - work done overlaps dotnet/arcade#2031 and dotnet/arcade#2033
- basically, iterate enough to avoid C# compilation errors in aspnet/AspNetCore repo when using `--all`

(probably easiest to review individual commits)

dougbu commented:
/fyi @pakrym @JunTaoLuo

dougbu commented on src/Microsoft.Cci.Extensions/Extensions/TypeExtensions.cs:
```diff
@@ -490,24 +490,27 @@ public static bool IsPropertyOrEventAccessor(this IMethodDefinition method)
-            foreach (var p in methodDefinition.ContainingTypeDefinition.Properties)
+            // Cannot use MemberHelper.IsAdder(...) and similar due to their TypeMemberVisibility.Public restriction.
+            var name = methodDefinition.GetNameWithoutExplicitType();
+            if (name.StartsWith("add_"))
```
Would be nice to use info from MethodSemantics table but CCI doesn't expose that information in a clear manner. But, if I'm missing something, let me know.

ericstj replied:
Won't this change incorrectly represent explicitly named methods?  I think the method prefixes are just a convention, there is actual metadata that represents properties and events.  Here's a [sample](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1RhYAdsAQAoBOA9gA4CmrwAnhQIBzbsADcBAM5jJAXyo1F1APQqRYggAsAhlILB22nQDduBAK70wARwvn6OgLbmY9g0fpH2Z1hHY6MAzCyrQEMOwWAEYQ5qLAAPpsXLwCALIAFACUQsgA7AQADPKhoWoEXsDmUs4OtVAEURZMQQBmrbzcjAQ6rCHYNGHIACzSYkkcPHz8GRHRsQQmOhD2OeQKA0qb1PgE3GbdAKIHwAASOvQwsawAPMddwACCfVIAfAQAyuwu993kPTAYOsCKxuE4fNxgRtBmV1JVqrUKoimi0wO1Ot1ev1BrsRgCYAkvj8ThlfmcLldeDcAKr0XSXWIwQ4ADwAxtxOMAwOx6GTnsI3sYGbw1gRoTRcaNQeCzITvtwyaSTudhbdafTKUy2RyuTy+S93hrrqKNnIgA=).  According to ECMA spec 335, I.8.11.3

> The CTS places no restrictions on the set of methods associated with a property, their names, or their
usage. 

Same is true for events.  I think you'll need to keep the code here that was considering property and event definitions.  What were you trying to fix with these changes?

dougbu replied:
This change addresses dotnet/arcade#2066. `Events` and `Properties` only contain members added in the containing class. However, in the case where a class adds an `interface` with some members implemented in a base class, the class contains only accessor methods (explicit implementations that call use the base event or property) and not the associated events or properties. Note the base class does _not_ implement the interface, just happens to have some members that align.

I.10.4 "Naming patterns" in the CTS spec also says
> While the CTS does not dictate the naming of properties or events, the CLS does specify a pattern to be observed.

The names used here match those specified in that section. In any case, explicitly-named methods can't have `IsSpecialName == true`.

ericstj replied:
I see.  Walking the type hierarchy would be pretty onerous here. I'm ok with the hueristic.  I agree that it there are other parts of the spec that say this is required for properties and events.

dougbu commented on src/Microsoft.Cci.Extensions/Filters/ExcludeAttributesFilter.cs:
```diff
@@ -7,22 +7,26 @@
         }
 
+        // Exists to avoid breaking change in removing includeForwardedTypes constructor parameters. This is a bit
+        // less ugly than suppressing IDE0060.
```
If it's possible to be sure nothing outside this repo, could take the breaking change. Anyone know?

ericstj replied:
@safern only recently added that.  I don't expect CCI.Extensions to be used much outside this repo, @terrajobst has some tools that use it, but I don't expect they'd be using this.  My opinion is just remove this parameter.  No reason this filter needs to know about forwarded types.

dougbu replied:
Done

safern commented on src/Microsoft.DotNet.GenAPI/Program.cs:
```diff
@@ -107,8 +122,14 @@ private static int Main(string[] args)
-                            writer = GetWriter(output, syntaxWriter);
+                            }
+
+                            var includeInternals = respectInternals.HasValue() &&
```
Can we avoid this duplication?

safern replied:
Ah right, the assembly/ies value is different on the if and the else. Can we make the attribute name a constant?

jaredpar commented on src/Microsoft.Cci.Extensions/Filters/ExcludeCompilerGeneratedCciFilter.cs:
```diff
@@ -0,0 +1,101 @@
+
+        public virtual bool Include(INamespaceDefinition ns)
+        {
+            return IsNotMarkedWithAttribute(ns) && ns.GetTypes(includeForwards: true).Any(Include);
```
> IsNotMarkedWithAttribute(ns [](start = 18, length = 28)

A namespace can't have an attribute associated with it hence I'm unsure what this check is doing.

jaredpar commented on src/Microsoft.Cci.Extensions/Filters/ExcludeCompilerGeneratedCciFilter.cs:
```diff
@@ -0,0 +1,101 @@
+            return IsNotMarkedWithAttribute(member);
+        }
+
+        public virtual bool Include(ICustomAttribute attribute)
```
>  public virtual bool Include(ICustomAttribute attribute) [](start = 6, length = 57)

Is this filtering custom attribute declarations or usages?

jaredpar replied:
Ah it looks like usages

---
In reply to: [363481250](https://github.com/dotnet/arcade/pull/4585#discussion_r363481250) [](ancestors = 363481250)

dougbu replied:
Yup, can this comment be resolved?

jaredpar replied:
Yes

dougbu commented:
ping @safern 

@ericstj @jaredpar is this ready to go?

ericstj commented:
I'll take another full look.

@dougbu can you make sure you test this package in dotnet/runtime repo before committing?  We need to test 1) library build, 2) /t:GenerateReferenceSouce from a couple projects: I'd suggest System.Runtime, System.Text.Json, and a few others.  @safern do you think you can help provide some suggestions for testing?

jaredpar commented:
@dougbu 

The impact of the change on reference assemblies looks good to me and lines up with what I'd expect in a reference assembly. I'm not very familiar with this project though hence I'm hesitant to "sign off" on the totality of the changes.

ericstj commented on src/Microsoft.Cci.Extensions/Extensions/TypeExtensions.cs:
```diff
@@ -490,24 +490,27 @@ public static bool IsPropertyOrEventAccessor(this IMethodDefinition method)
+            {
+                return AccessorType.EventRemover;
+            }
+            else if (name.StartsWith("set_"))
```
nit: get and set will be more common than add / remove.  Can you order the if/else as get, set, add, remove?

safern commented:
> @safern do you think you can help provide some suggestions for testing?

Those assemblies would be good tests. 

Also, after those changes running APICompat on those projects if there are significant changes we need to take on those reference assemblies. 

When I did the nullable changes, I also had a dummy assembly with a lot of edge cases and ran it against that.

ericstj commented on src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs:
```diff
@@ -33,11 +33,23 @@ private void WriteMethodDefinition(IMethodDefinition method)
-                if (!method.IsExplicitInterfaceMethod()) WriteVisibility(method.Visibility);
+                if (method.IsMethodUnsafe())
+                {
+                    WriteKeyword("unsafe");
```
It looks to me like `WriteInterfaceMethodModifiers` is supposed to write modifiers that are common to both methods and interfaces methods.  Can you instead just move unsafe check out of `WriteMethodModifiers` and into `WriteInterfaceMethodModifiers`?  Granted this would change the ordering of the modifiers.

dougbu replied:
Unfortunately, can't do this because `WriteMethodModifiers(...)` is also called to generate the modifiers for events and properties. Both can be declared `unsafe`.

ericstj commented on src/Microsoft.DotNet.GenAPI/Program.cs:
```diff
@@ -244,24 +273,38 @@ private static string GetFilename(IAssembly assembly, WriterType writer, SyntaxW
+            else if (includeInternals)
+            {
+                includeFilter = new InternalsAndPublicCciFilter(excludeAttributes: apiOnly, includeForwardedTypes);
             }
```
Do you want to have any friendly error messages to indicate incompatible or ignored options?

ericstj replied:
I see your perspective.  These are ordered in a way that all > public+internals > public.  I think list is straight forward enough that it doesn't require a warning.

ericstj commented on src/Microsoft.Cci.Extensions/Filters/ExcludeCompilerGeneratedCciFilter.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <item>Is specific to <see cref="T:System.Runtime.CompilerServices.CompilerGeneratedAttribute"/>.</item>
+    /// <item>Includes property and event accessors despite annotations.</item>
+    /// <item>Excludes leftover <see cref="T:System.Runtime.CompilerServices.CompilerGeneratedAttribute"/>s.</item>
+    /// <item>Adds parameter <see langword="null"/> checks.</item>
```
Just make them consistent, there is no need for two filters to differ in their null checking if they are called by the same code.

ericstj commented on src/Microsoft.Cci.Extensions/Filters/InternalsAndPublicCciFilter.cs:
```diff
@@ -0,0 +1,181 @@
+            // If a type is abstract and has an internal or public constructor, it must expose all abstract members.
+            if (containingType.IsAbstract && member.IsAbstract())
+            {
+                foreach (var method in containingType.Methods)
```
You could add an optional param to https://github.com/dotnet/arcade/blob/ef4b288de587f0b86e65b3950e9095f714808ade/src/Microsoft.Cci.Extensions/Extensions/TypeExtensions.cs#L721 or create a new extension method?

ericstj commented on src/Microsoft.Cci.Extensions/Filters/InternalsAndPublicCciFilter.cs:
```diff
@@ -0,0 +1,181 @@
+                }
+            }
+
+            // Include explicit interface implementations. (Inspired by MemberHelper.IsVisibleOutsideAssembly(...).)
```
I'd create MemberExtensions.cs in https://github.com/dotnet/arcade/tree/master/src/Microsoft.Cci.Extensions/Extensions, similar to what's already done with TypeExtensions

dougbu commented on src/Microsoft.Cci.Extensions/Filters/ExcludeCompilerGeneratedCciFilter.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <item>Includes property and event accessors despite annotations.</item>
+    /// <item>Excludes leftover <see cref="T:System.Runtime.CompilerServices.CompilerGeneratedAttribute"/>s.</item>
+    /// <item>Adds parameter <see langword="null"/> checks.</item>
+    /// <item>Avoids <see cref="NullReferenceException"/>s when handling <see cref="FakeCustomAttribute"/>s.</item>
```
`AttributeMarkedFilter.Include(ICustomAttribute)` unconditionally returns `true` because it doesn't special-case some members marked with the chosen attribute i.e. there's no "leftovers". The `FakeCustomAttribute` case isn't relevant there.  I'm just removing this comment.

ericstj commented on src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs:
```diff
@@ -340,6 +355,11 @@ private void WritePrivateConstructor(ITypeDefinition type)
                 return;
 
             WriteVisibility(TypeMemberVisibility.Assembly);
```
I see.  I believe it needed to be internal so that we can write derived classes in the same assembly.  If it's private a derived class would need to call the "right" constructor, and I'm not sure GenAPI does this.

dougbu replied:
Is the concern manual additions to ref/ assemblies? This constructor won't be called from generated code. `GetBaseConstructorForCall(...)` only discovers unfiltered constructors of the base class -- `public` and `protected` for your cases I believe. See https://github.com/dotnet/arcade/pull/4585/files#diff-550675841714877a926fc473e10e9ba2R419

See also https://github.com/dotnet/arcade/pull/4585/files#diff-cbc92831d638d8671626f77d630ef11bR133-R134

Bottom line, I'd like to make this constructor `private`. One alternative is something like creating a fake `internal` constructor and using `private` if `_filter.Include(fakeInternal)` and `internal` otherwise. Another option might be to remove this method and the `CSDeclarationWriter.DummyInternalConstructor` special cases, then change `CSharpWriter.Visit(ITypeDefinition)` to write _all_ constructors.

ericstj replied:
I'm not raising a concern, just trying to recall why this is doing what its doing.  I believe there are cases with derived classes.  I could be wrong.  Don't write all constructors, just test to vet your change.

dougbu commented:
FYI I'm nearly ready to update this PR to resolve the conflict, address review comments, and fix a couple of problems hit in dotnet/aspnetcore. That repo (with changes to use `--respect-internals` and so on) now generates clean ref/ code and passes ApiCompat checks in both directions with **no** errors. I'm testing in dotnet/runtime now&hellip;

> When I did the nullable changes, I also had a dummy assembly with a lot of edge cases and ran it against that.

This was a good idea but not something I did in this case. The dotnet/aspnetcore repo contains all of the edge cases I was trying to fix i.e. I wasn't adding support for new C# features. Of course, please let me know of any corner cases I should double-check; hopefully I can find them where I've already tested.

dougbu commented:
Generating a `private` constructor when no other constructor exists doesn't always work in dotnet/runtime because some ref/ code calls filtered-out `internal` constructors. **Which of the following do you prefer?**

1. Add a command-line option controlling the "default" generated constructor, switching it between `private` and `internal`
2. Base the choice between `private` and `internal` on `--respect-internals`
3. Base the choice between `private` and `internal` on whether a filtered-out `internal` constructor exists i.e. the code could check the constructor collection again in `WritePrivateConstructor(...)` (without the filter)
4. Add a list of `internal` constructors to the *.Manual.cs files. This list shouldn't be incredibly long; I hit `is inaccessible due to its protection level` problems in only a few cases e.g. for the `JsonNode()` constructor
5. Add a `UnionFilter` that prevents exclusion of a few docIDs from generation as well as a corresponding command-line switch then use that in dotnet/runtime e.g. through `$(GenAPIAdditionalParameters)`

----

When testing with `dotnet msbuild /t:GenerateReferenceSource` or `msbuild /t:GenerateReferenceSource` in the dotnet/runtime repo, GenApi emits errors about resolving 'System.Private.CoreLib` in a few src directories, including src\libraries\System.Runtime\src\ e.g.

```
  Unable to resolve assembly 'Assembly(Name=System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)' referenced by 'Assembly(Name=System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)'.
```

The above occurs despite having built the libraries subcategory completely before starting. **What am I missing?**

----

Separately from the above, I see some differences in the generation for `ExpandoObject`:

``` diff
-        bool System.Collections.Generic.IDictionary<string, object?>.TryGetValue(string key, out object? value) { throw null; }
-        System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, object?>> System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>.GetEnumerator() { throw null; }
+        bool System.Collections.Generic.IDictionary<string, object?>.TryGetValue(string key, out object value) { throw null; }
+        System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, object>> System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>.GetEnumerator() { throw null; }
```

Unlike other cases where the nullability generation has changed, the above code was correct _before_ I ran GenApi. Before I get deep into debugging the change, **is this a known issue that you've been working around?**

dougbu commented:
:up::date: to current state of my branch.

My updated ref/ code for dotnet/runtime is visible at https://github.com/dotnet/runtime/compare/master...dougbu:dougbu/local.GenApi?expand=1

dougbu commented on src/Microsoft.DotNet.ApiCompat/build/Microsoft.DotNet.ApiCompat.targets:
```diff
@@ -32,7 +32,7 @@
 
     <MatchingRefApiCompatBaseline Condition="!Exists('$(MatchingRefApiCompatBaseline)')">$(MSBuildProjectDirectory)\MatchingRefApiCompatBaseline.$(_apiCompatTargetSuffix).txt</MatchingRefApiCompatBaseline>
-    <MatchingRefApiCompatBaseline Condition="'$(BaselineAllMatchingRefApiCompatError)' != 'true' and !Exists('$(MatchingRefApiCompatBaseline)')">$(MSBuildProjectDirectory)\MatchingRefApiCompatBaseline.txt</MatchingRefApiCompatBaseline>
+    <MatchingRefApiCompatBaseline Condition="!Exists('$(MatchingRefApiCompatBaseline)')">$(MSBuildProjectDirectory)\MatchingRefApiCompatBaseline.txt</MatchingRefApiCompatBaseline>
```
As mentioned in commit comments, the previous `Condition` meant baselines were read and written from different files by default

ericstj replied:
I suspect this was done to make the default behavior to write to TargetFramework-specific baseline files, though I am OK with this change.

dougbu commented on src/Microsoft.DotNet.ApiCompat/build/Microsoft.DotNet.ApiCompat.targets:
```diff
@@ -68,8 +68,8 @@
       <ApiCompatArgs Condition="'$(ApiCompatExcludeAttributeList)' != ''">$(ApiCompatArgs) --exclude-attributes "$(ApiCompatExcludeAttributeList)"</ApiCompatArgs>
       <ApiCompatArgs Condition="'$(ApiCompatEnforceOptionalRules)' == 'true'">$(ApiCompatArgs) --enforce-optional-rules</ApiCompatArgs>
       <ApiCompatArgs Condition="'$(BaselineAllAPICompatError)' != 'true' and Exists('$(ApiCompatBaseline)')">$(ApiCompatArgs) --baseline "$(ApiCompatBaseline)"</ApiCompatArgs>
-      <!-- Must be last option. -->
```
This was a lie

ericstj replied:
Probably was true before https://github.com/dotnet/arcade/commit/c380971edbf944301a3b8bc9bd89f4606b4dedc0.

dougbu commented on src/Microsoft.DotNet.ApiCompat/Program.cs:
```diff
@@ -120,8 +125,21 @@ public static int Main(string[] args)
-                            excludeNonBrowsable.HasValue(), remapFile.Value(), !skipGroupByAssembly.HasValue(), leftOperandValue, rightOperandValue, excludeAttributes.Value());
+                        var includeInternals = respectInternals.HasValue() &&
+                            contractAssemblies.Any(assembly => assembly.Attributes.HasAttributeOfType(
+                                "System.Runtime.CompilerServices.InternalsVisibleToAttribute"));
```
No `const` here because the `string` is used only in this one place. Any objections?

ericstj commented:
> 2. Base the choice between private and internal on --respect-internals

I'm OK with this.

> GenApi emits errors about resolving 'System.Private.CoreLib`

Do you see System.Private.CoreLib being passed in to GenAPI?  If not, try adding `DependsOnTargets="ResolveReferences"` here: https://github.com/dotnet/runtime/blob/86a086377bd63cbe678b6cba5498e7ab4fd9b57d/src/libraries/Directory.Build.targets#L117
Not entirely sure why it doesn't have it /cc @safern @Anipik 

> I see some differences in the generation for ExpandoObject

Yes, we're working around nullable annotations on explicit interface implementations.  https://github.com/dotnet/arcade/issues/4722

ericstj commented on src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs:
```diff
@@ -339,7 +354,12 @@ private void WritePrivateConstructor(ITypeDefinition type)
                 return;
 
-            WriteVisibility(TypeMemberVisibility.Assembly);
+            WriteVisibility(TypeMemberVisibility.Private);
```
I replied.  Can you please record the reasoning for the WritePrivateConstructor in a comment when making your change?

pgrawehr commented:
Could somebody please take a look at the related PR #4649? I think it's an easy but important fix.

ericstj commented:
@dougbu let me know if you're waiting on anything here.  I can take another look if you have specific places for me to check.  I was under the impression you were going to make some changes.

dougbu commented:
@ericstj thanks. Right now, I'm just resolving an ApiCompat issue troubling the dotnet/aspnetcore repo: The checks work fine in one direction but fail due to "removed" `[CompilerGenerated]` attributes. Basically, I'm looking into why filters we apply (specifically the one that cleans out that attribute and many things where it's used) are actually used in only one direction.

Nothing other than that, redoing dotnet/runtime testing, and (of course) other urgent work is holding me up.

dougbu commented:
@ericstj adding the target dependency doesn't make errors go away but it does change them. Now see e.g.
```
C:\dd\dnx\runtime\eng\referenceFromRuntime.targets(112,5): error : Could not resolve ReferenceFromRuntime item(s) 'System.Private.CoreLib' from 'C:\dd\dnx\runtime\src\libraries\restore\runtime\runtime.depproj'. [C:\dd\dnx\runtime\src\libraries\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj]
```

FYI I'm executing
```
> git clean -dfx
> build.cmd -subsetCategory coreclr-libraries -runtimeConfiguration Release
# This hits about 14 errors but deep in the coreclr phase of the build. Not enough to prevent libraries projects from building.
> dir .\src\libraries\*\src\*.csproj  |% {echo $_.FullName; dotnet msbuild /t:GenerateReferenceSource $_.FullName}
> dir .\src\libraries\*\ref\*.cs |? {$_.Length -lt 480} |% {git restore $_.FullName} || {rm $_.FullName}
# This restores files left truncated when GenApi fails.
```

Do the `dotnet msbuild` commands need additional options?

dougbu commented:
Today's result of the above commands is captured at https://github.com/dotnet/runtime/compare/master...dougbu:dougbu/local.GenApi?expand=1 I believe none of what's shown there is unexpected but don't have full context.

dougbu commented:
:up::date: to add a few more (four?) fixes. Believe I'm done iterating because all tests in dotnet/aspnetcore and dotnet/runtime seem pretty clean.

Generating `internal`s in dotnet/aspnetcore has allowed me to remove all but a handful of *.Manual.cs files and GenApi exclusions. And, my fork of our 'release/3.1' branch is able to execute ApiCompat in both directions without error.

The result of regenerating in dotnet/runtime (https://github.com/dotnet/runtime/compare/master...dougbu:dougbu/local.GenApi?expand=1) looks messy but all of it seems explainable by
- a fair number of manual changes to the generated files that I haven't tried to replicate
- real API changes since last GenApi use
- attribute exclusions due to eng\DefaultGenApiDocIds.txt changes since last GenApi use
- recent (or recent-ish) changes in generation for nullability

I believe the root cause of almost all build errors (missing and duplicate members when executing `.\build.cmd -subsetCategory libraries -runtimeConfiguration Release`) is manual changes. Of course, I may be missing something.

safern commented:
I skimmed through your changes. 

1. The EditorBrowsableAttribute, ObsoleAttribute changes should be ignored. We haven't enforced people to add those in the sources, so that's why they're removed.

2. Some of the API additions are either, in a Manual file already and we need to remove them from there. I believe we added support already in runtime to provide a list of APIs we want to ignore when running GenAPI.

3. Some other API additions are intentionally not exposed, there are some special cases, like in System.Runtime, where the APIs are made public in System.Private.CoreLib, but since we use them in some implementation in other assemblies, we don't wanted to use InternalsVisibleTo, so we made them public, not expose them, and then suppress the ApiCompat errors with a baseline files.

What I saw, makes sense, and I didn't see any concerning changes from what I skimmed. I think for runtime, we would need to update the ref assemblies if needed, but I would do it by small pieces so that the diff is not giant.

dougbu commented:
> I skimmed through your changes.

Thanks @safern❕ Good to know I didn't miss anything important.

@ericstj @safern is this PR ready to be squashed and merged? (hmm, what's the preference for PRs w/ multiple small pieces included in this repo?)

ericstj commented on src/Microsoft.DotNet.ApiCompat/build/Microsoft.DotNet.ApiCompat.targets:
```diff
@@ -121,15 +122,16 @@
-      <MatchingRefApiCompatArgs>$(MatchingRefApiCompatArgs) "$(ImplemetnationAssemblyAsContract)"</MatchingRefApiCompatArgs>
-      <MatchingRefApiCompatArgs>$(MatchingRefApiCompatArgs) --contract-depends "@(_ContractDependencyDirectories, ','),"</MatchingRefApiCompatArgs>
+      <MatchingRefApiCompatArgs>$(MatchingRefApiCompatArgs) "$(ImplementationAssemblyAsContract)"</MatchingRefApiCompatArgs>
+      <MatchingRefApiCompatArgs>$(MatchingRefApiCompatArgs) --contract-depends "$(IntermediateOutputPath),@(_ContractDependencyDirectories, ','),"</MatchingRefApiCompatArgs>
```
Why do you need IntermediateOutputPath here?

dougbu replied:
Needed to ensure this command is the `true` inverse of the forward check. Without it, assemblies copied into obj/ weren't resolved when doing this reverse check.

See https://github.com/dotnet/arcade/pull/4585/files#diff-3c3f2c92eeec9b69dac2deccfac46603R71

ericstj replied:
I understand the purpose of this commandline.  I'm asking in what case are assemblies copied to `obj` but not passed in as `ReferencePath` to the compiler?  We already have `_ContractDependencyDirectories` which should serve this purpose, if that's not working correctly then I'd like to understand why.

dougbu replied:
:ok: I'll retest dotnet/aspnetcore w/ the latest code and see if the intermediate output path is duplicated in `_ContractDependencyDirectories`. Will report back here later today.

ericstj replied:
I'm not expecting _ContractDependencyDirectories to contain the intermediate path, I'm just expecting that everything that might have been copied to that path was also passed to the compiler as ReferencePath and thus in _ContractDependencyDirectories.  If you are finding things in intermediate and not ReferencePath then I'd wonder how they got there.

dougbu replied:
I made this change while attempting to figure out why forward checks worked but reverse aka matching checks did not. Made the reverse `--contract-depends` consistent with the `--impl-dirs` but my [other change here](https://github.com/dotnet/arcade/pull/4585/files#diff-3c3f2c92eeec9b69dac2deccfac46603R116) actually fixed the problem.

# TL;DR

Consistency here is good for projects using the Razor SDK but we could remove `$(IntermediateOutputPath)` from `--impl-dirs` / `--contract-depends` as long as we're sure we don't need ApiCompat for projects using that SDK and there aren't other similar cases. See questions around this at the bottom.

# Does including `$(IntermediateOutputPath)` ever matter?

I can't find any obj/ directory in dotnet/aspnetcore or dotnet/runtime containing more than one DLL except where the Razor SDK generated an additional Foo.Views.dll assembly and none of those projects have associated ref/ assemblies let alone ApiCompat checks. This change didn't actually fix anything in either repo; our repos don't currently justify including `$(IntermediateOutputPath)` in `--impl-dirs` (for forward checks).

## Razor SDK

Digging into one of the projects that uses the Razor SDK, `@(ReferencePath)` isn't updated when the project executes -- likely because Foo.Views.dll depends on Foo.dll and not the other way around. The Razor SDK mucks with `@(RazorReferencePath)` to make this work, not `@(ReferencePath)`. Therefore `$(IntermediateOutputPath)` **does** need to be included in `--impl-dirs` if we run ApiCompat checks on *.Views.dll assemblies.

## History

Including `$(IntermediateOutputPath)` predates the Razor SDK. I traced [the relevant line](https://github.com/dotnet/arcade/pull/4585/files#diff-3c3f2c92eeec9b69dac2deccfac46603R71) all the way back to https://github.com/dotnet/buildtools/commit/752c0f0bedb0fc735a37dd4719ce2c7937129a78#diff-d32defe7101451053df5beb560c68ce3R30 -- when @venkat-raman251 introduced the original ApiCompat.targets file.  The `ValidateApiCompatForSrc` target originally used `$(OutputPath)` and depended on the `CopyFilesToOutputDirectory` target. The only related changes since have been switching to `$(IntermediateOutputPath)` (and removing the target dependency) and updating the command-line argument names.

Perhaps when @weshaggard switched `$(ValidateApiCompatForSrc)` to not rely on `CopyFilesToOutputDirectory`, he could have removed `$(OutputPath)` instead of switching to `$(IntermediateOutputPath)`

## Questions

1. @jaredpar @weshaggard any insights here i.e. do you know of non-Razor SDK cases where assemblies land in `$(IntermediateOutputPath)` but aren't included in `@(ReferencePath)`?
2. If we ignore the Razor SDK, are there even cases where or more than one assembly lands in an obj/ directory?
3. Anyone else we need to ask about these cases and potential cases?
4. @Pilchie for our 'release/3.1' branch, we originally wanted to also check public API compatibility against assemblies from 3.1.0 packages. Is that still a priority? Would we check the *.Views.dll assemblies in this?

Pilchie replied:
Basically, the idea was to prevent ourselves from accidentally changing the surface area in a patch release.  I think that's not an issue because the check against the ref assemblies would fail?  Assuming that's true, I don't think actually comparing directly against the 3.1.0 ref assemblies is important.

dougbu replied:
> I think that's not an issue because the check against the ref assemblies would fail?

Not quite. We only have ref/ assemblies for shared framework assemblies, nowhere near all assemblies we ship.

I do agree we're safe w.r.t. the shared framework as long as we're diligent on changes to the ref/ projects or (perhaps) check against a published 3.1.x Microsoft.AspNetCore.App.Ref package (whichever was the earliest complete one).

Pilchie replied:
But things that aren't in the shared framework *are* uniquely addressable by package version, so it's less of an issue if surface area does change there (though not ideal).

dougbu replied:
Good point. We're therefore fine ignoring what the Razor SDK does.

Back to: Are there other ">1 assembly in obj/" cases that we _should_ care about?

Pilchie replied:
I don't know of any.

dougbu replied:
@ericstj @jaredpar do you know of other ">1 assembly in obj/" cases?

dougbu replied:
@ericst assuming your and @jaredpar's silence means there aren't other cases we need to worry about, do you want me to remove `$(IntermediateOutputPath),` here and in the `ValidateApiCompatForSrc` target before merging?

ericstj replied:
I don't know of any case, which is why I raised this question to begin with.

> do you want me to remove $(IntermediateOutputPath), here and in the ValidateApiCompatForSrc target before merging?

Yeah, let's not add this in, and you can clean up the other target where it was unnecessary.  If we rely on this somewhere I'd really like to see that before we include it.

Thanks!

dougbu replied:
Removed.

safern commented on src/Microsoft.Cci.Extensions/Filters/ExcludeAttributesFilter.cs:
```diff
@@ -7,28 +7,29 @@
         private readonly HashSet<string> _attributeDocIds;
 
-        public ExcludeAttributesFilter(IEnumerable<string> attributeDocIds, bool includeForwardedTypes = false)
-            : base(excludeAttributes: false, includeForwardedTypes: includeForwardedTypes)
```
Is this going to break the follow type forwards feature?

dougbu replied:
No, this filter isn't used on its own and has no reason to double-check that a namespace is empty.

safern commented on src/Microsoft.Cci.Extensions/Filters/InternalsAndPublicCciFilter.cs:
```diff
@@ -0,0 +1,120 @@
+        public virtual bool Include(ITypeDefinitionMember member)
+        {
+            // Include based on member visibility (simple cases).
+            switch (member.Visibility)
```
Why can't we use the helper methods to figure it's visibility?

dougbu replied:
Good catch. I created `MemberExtensions.IsVisibleToFriendAssemblies(ITypeDefinitionMember)` after writing this and should have removed this `switch` when I started using that method. Small redundancy but I'll remove when I rebase this branch.

dougbu replied:
Oh, wait. The problem with `MemberExtensions.IsVisibleToFriendAssemblies(ITypeDefinitionMember member)` is its check of `member.ContainingTypeDefinition.IsSealed` in the `protected internal` and `internal` cases. That check makes sense for other uses of the method but not here; we may need sealed methods for calls from implementation assemblies with IVT to this one.

That said, I can reduce the `switch` to an `if` that just checks `TypeMemberVisibility.Family` and `...FamilyAndAssembly`.

dougbu replied:
Done

safern replied:
Awesome. Thanks.

dougbu commented:
@safern anything other than the https://github.com/dotnet/arcade/pull/4585#pullrequestreview-383262214 discussion blocking your approval?

dougbu commented:
FYI the last update included only the `$(IntermediateOutputPath)` removal and a final rebase. Will merge when the checks complete and I've run local tests in dotnet/aspnetcore and dotnet/runtime.
